src/main/java/org/testng/TestRunner.java

```diff
@@ -474,7 +474,7 @@ private void initMethods() {
     //
     for (ITestClass tc : m_classMap.values()) {
       fixMethodsWithClass(tc.getTestMethods(), tc, testMethods);
-      fixMethodsWithClass(tc.getBeforeClassMethods(), tc, beforeClassMethods);
+      fixMethodsWithClass(buildBeforeClassMethods(tc.getBeforeClassConfig()), tc, beforeClassMethods);
```

Comment:
Also it looks like after the methods have been fixed with the corresponding Objects, we can actually clean-up this `beforeClassConfig` map.